reformatted multiplex postprocessing

ngreenwald opened the pull request:
This PR brings in the new functionality vanvalenlab/deepcell-tf#413 for better formatting of model output, and also adapts the existing post-processing functions to handle it. 

- Add format_output_multiplex() to take the list of model outputs, and convert them into a nicely structured dict
- Add deep_watershed_subcellular() to take the formatted model output, and pass either the nuclear predictions, the whole-cell predictions, or both, to deep_watershed_mibi()
- Change deep_watershed_mibi() to expect a dictionary of model outputs, rather than a list
- Pull out additional hard-coded parameters in deep_watershed_mibi and add them as default function arguments to enable them to be changed if desired
- Wrap all of these in a kwargs dict and hide them in high-level functions

willgraf commented on deepcell_toolbox/deep_watershed.py:
```diff
@@ -101,55 +101,81 @@ def deep_watershed(outputs,
+    cell_prediction_batch = model_output[cell_model]
+    maxima_prediction_batch = model_output[maxima_model]
+
+    if len(cell_prediction_batch.shape) != 4:
+        raise ValueError('Model output must be of length 4. The cell_prediction model '
+                         'provided was of shape {}'.format(cell_prediction_batch.shape))
```
Maybe a loop since the text is so similar?

```
zipped = zip(['cell_prediction', 'maxima_prediction'], (cell_prediction_batch, maxima_prediction_batch))
for name, arr in zipped:
    if len(arr.shape) !=4:
        raise ValueError(...)
```

willgraf commented on deepcell_toolbox/deep_watershed.py:
```diff
@@ -161,5 +187,88 @@ def deep_watershed_mibi(outputs,
+    if len(output_list) != 8:
+        raise ValueError('output_list was length {}, expecting length 8'.format(len(output_list)))
+
+    formatted_dict = {'whole-cell': {'inner-distance': output_list[0],
+                                     'outer-distance': output_list[1],
+                                     'fgbg-fg': output_list[2][..., :1],
+                                     'pixelwise-interior': output_list[3][..., 1:2]},
+                      'nuclear': {'inner-distance': output_list[4],
+                                  'outer-distance': output_list[5],
+                                  'fgbg-fg': output_list[6][..., :1],
+                                  'pixelwise-interior': output_list[7][..., 1:2]}
+                      }
```
Also, just curious, what's up with the slicing in fgbg and pixelwise?

ngreenwald replied:
I did that so that each semantic head produces a single [batch x rows x cols x 1] array that can be used for watershed. We always use the foreground prediction from fgbg, and the interior prediction from pixelwise, for the postprocessing. That way they can be swapped in by specifying a different maxima_model name without any code changes

ngreenwald commented on deepcell_toolbox/deep_watershed.py:
```diff
@@ -101,55 +101,81 @@ def deep_watershed(outputs,
-            at the border.
+        min_distance (int): Minimum allowable distance between two cell maxima.
+        maxima_threshold (float): Threshold for the maxima prediction.
+        cell_threshold (float): Threshold for the cell prediction.
+        exclude_border (bool): Whether to include centroid detections at the border.
         small_objects_threshold (int): Removes objects smaller than this size.
+        cell_model: semantic head to use to predict location of cells
```
There's two requirements for the actual watershed function. Something to use to calculate maxes, and something to use to calculate the background/foreground for watershedding over. 

I've used cell prediction here to mean the thing being thresholded that determines foreground/background, i.e. where cells are located in the image. Open to other names. I've kept it agnostic to the specific semantic head being fed in to that thresholding step

willgraf replied:
That is great. I'm trying to think of something that's more general than "cell"...maybe "object" or "target" or something?

ngreenwald replied:
Yeah, on the angelolab side i've been using interior and maxima, but then for this PR I thought maybe cell was actually better?

ngreenwald replied:
I like object or shape?

willgraf replied:
I honestly like "interior" the best, but I also like "object" and "shape". You can pick your favorite, but I do think we should change it to more general than "cell".

ngreenwald commented on deepcell_toolbox/deep_watershed.py:
```diff
@@ -101,55 +101,81 @@ def deep_watershed(outputs,
+        raise ValueError('maxima_model must be one of {}, got {}'.format(valid_model_names,
+                                                                         maxima_model))
+
+    cell_prediction_batch = model_output[cell_model]
+    maxima_prediction_batch = model_output[maxima_model]
+
+    if len(cell_prediction_batch.shape) != 4:
+        raise ValueError('Model output must be of length 4. The cell_prediction model '
+                         'provided was of shape {}'.format(cell_prediction_batch.shape))
+
+    if len(maxima_prediction_batch.shape) != 4:
+        raise ValueError('Model output must be of length 4. The maxima_prediction model '
+                         'provided was of shape {}'.format(maxima_prediction_batch.shape))
```
okay, so what would we call the current batch derived from this array?

willgraf replied:
It could just be the singular form (`cell_prediction`) or a short variable name (`cp`). The short version should probably only be used if its used just a couple times, if the scoped-in code is a little long than use the former.

ngreenwald commented:
Okay, I switched back to 'interior' naming scheme. 
I added more tests for the newly created ValueError checks in the deep_watershed_mibi function. 
I committed your formatting suggestions
I shortened the variable names for the batches

willgraf commented on deepcell_toolbox/deep_watershed.py:
```diff
@@ -101,65 +101,186 @@ def deep_watershed(outputs,
 
         # Remove small objects
         label_image = remove_small_objects(label_image, min_size=small_objects_threshold)
+        print(label_image.shape)
```
```suggestion
```
